File: src/main/java/net/sjrx/intellij/plugins/systemdunitfiles/completion/UnitFileKeyCompletionContributor.java
package net.sjrx.intellij.plugins.systemdunitfiles.completion;

import com.intellij.codeInsight.completion.*;
import com.intellij.codeInsight.lookup.LookupElement;
import com.intellij.codeInsight.lookup.LookupElementBuilder;
import com.intellij.openapi.editor.EditorModificationUtil;
import com.intellij.openapi.util.text.StringUtil;
import com.intellij.patterns.PlatformPatterns;
import com.intellij.psi.PsiElement;
import com.intellij.psi.util.PsiTreeUtil;
import com.intellij.util.ProcessingContext;
import net.sjrx.intellij.plugins.systemdunitfiles.UnitFileIcon;
import net.sjrx.intellij.plugins.systemdunitfiles.UnitFileLanguage;
import net.sjrx.intellij.plugins.systemdunitfiles.generated.UnitFileElementTypeHolder;
import net.sjrx.intellij.plugins.systemdunitfiles.psi.UnitFilePropertyType;
import net.sjrx.intellij.plugins.systemdunitfiles.psi.UnitFileSectionGroups;
import net.sjrx.intellij.plugins.systemdunitfiles.semanticdata.SemanticDataRepository;
import org.jetbrains.annotations.NotNull;

import java.util.Set;
import java.util.stream.Collectors;

/**
 * Completion Contributor for keywords in a section.
 */
public class UnitFileKeyCompletionContributor extends CompletionContributor {


  private SemanticDataRepository sdr = SemanticDataRepository.getInstance();

  /**
   * Default constructor.
   */
  public UnitFileKeyCompletionContributor() {
    extend(CompletionType.BASIC, PlatformPatterns.psiElement(UnitFileElementTypeHolder.KEY).withLanguage(UnitFileLanguage.INSTANCE),
      new CompletionProvider<CompletionParameters>() {

        @Override
        protected void addCompletions(@NotNull CompletionParameters parameters,
                                      @NotNull ProcessingContext context,
                                      @NotNull CompletionResultSet resultSet) {
          UnitFileSectionGroups section = PsiTreeUtil.getParentOfType(parameters.getOriginalPosition(), UnitFileSectionGroups.class);
          if (section == null ) section = PsiTreeUtil.getParentOfType(parameters.getPosition(), UnitFileSectionGroups.class);
          if (section == null) return;

          String sectionName = section.getSectionName();
          Set<String> definedKeys = section.getPropertyList().stream().map(UnitFilePropertyType::getKey).collect(Collectors.toSet());

          for (String keyword : sdr.getDocumentedKeywordsInSection(sectionName)) {
            if (definedKeys.contains(keyword)) continue;
            LookupElementBuilder builder = LookupElementBuilder
                    .create(keyword)
                    .withInsertHandler(new KeyInsertHandler())
                    .withPresentableText(keyword)
                    .withIcon(UnitFileIcon.FILE).appendTailText("(Option)", true);

            resultSet.addElement(builder);
          }
        }
      }
    );
  }

  @Override
  public void fillCompletionVariants(@NotNull CompletionParameters parameters, @NotNull CompletionResultSet result) {
    super.fillCompletionVariants(parameters, result);
  }

  private static class KeyInsertHandler implements InsertHandler<LookupElement> {
    @Override
    public void handleInsert(@NotNull InsertionContext context, @NotNull LookupElement item) {
      PsiElement element = context.getFile().findElementAt(context.getStartOffset());
      if (element == null) return;
      if (element.getNode().getElementType() == UnitFileElementTypeHolder.KEY) {
        PsiElement next = PsiTreeUtil.nextVisibleLeaf(element);
        if (next != null && next.getNode().getElementType() == UnitFileElementTypeHolder.SEPARATOR) {
          return;
        }
      }

      if (StringUtil.containsChar(" =", context.getCompletionChar())) {
        context.setAddCompletionChar(false);
      }
      EditorModificationUtil.insertStringAtCaret(context.getEditor(), "=");
    }
  }
}

File: src/main/kotlin/net/sjrx/intellij/plugins/systemdunitfiles/completion/UnitFileKeyCompletionContributor.kt
package net.sjrx.intellij.plugins.systemdunitfiles.completion

import com.intellij.codeInsight.completion.*
import com.intellij.codeInsight.lookup.LookupElement
import com.intellij.codeInsight.lookup.LookupElementBuilder
import com.intellij.openapi.editor.EditorModificationUtil
import com.intellij.openapi.util.text.StringUtil
import com.intellij.patterns.PlatformPatterns
import com.intellij.psi.util.PsiTreeUtil
import com.intellij.util.ProcessingContext
import net.sjrx.intellij.plugins.systemdunitfiles.UnitFileIcon
import net.sjrx.intellij.plugins.systemdunitfiles.UnitFileLanguage
import net.sjrx.intellij.plugins.systemdunitfiles.generated.UnitFileElementTypeHolder
import net.sjrx.intellij.plugins.systemdunitfiles.psi.UnitFileProperty
import net.sjrx.intellij.plugins.systemdunitfiles.psi.UnitFileSectionGroups
import net.sjrx.intellij.plugins.systemdunitfiles.semanticdata.SemanticDataRepository
import java.util.stream.Collectors

/**
 * Completion Contributor for keywords in a section.
 */
class UnitFileKeyCompletionContributor : CompletionContributor() {
  /**
   * Default constructor.
   */
  init {
    extend(CompletionType.BASIC, PlatformPatterns.psiElement(UnitFileElementTypeHolder.KEY).withLanguage(UnitFileLanguage.INSTANCE),
           object : CompletionProvider<CompletionParameters>() {
             override fun addCompletions(parameters: CompletionParameters,
                                         context: ProcessingContext,
                                         resultSet: CompletionResultSet) {
               var section = PsiTreeUtil.getParentOfType(parameters.originalPosition, UnitFileSectionGroups::class.java)
               if (section == null) section = PsiTreeUtil.getParentOfType(parameters.position, UnitFileSectionGroups::class.java)
               if (section == null) return
               val sectionName = section.sectionName
               val definedKeys = section.propertyList.stream().map { obj: UnitFileProperty -> obj.key }.collect(Collectors.toSet())
               val sdr = SemanticDataRepository.instance
               for (keyword in sdr.getDocumentedKeywordsInSection(sectionName)) {
                 if (definedKeys.contains(keyword)) continue
                 val deprecated = sdr.isDeprecated(sectionName, keyword)
                 val builder = LookupElementBuilder
                   .create(keyword)
                   .withInsertHandler(KeyInsertHandler())
                   .withPresentableText(keyword)
                   .withStrikeoutness(deprecated)
                   .withIcon(UnitFileIcon.FILE)
                 resultSet.addElement(builder)
               }
             }
           }
    )
  }

  private class KeyInsertHandler : InsertHandler<LookupElement> {
    override fun handleInsert(context: InsertionContext, item: LookupElement) {
      val element = context.file.findElementAt(context.startOffset) ?: return
      if (element.node.elementType === UnitFileElementTypeHolder.KEY) {
        val next = PsiTreeUtil.nextVisibleLeaf(element)
        if (next != null && next.node.elementType === UnitFileElementTypeHolder.SEPARATOR) {
          return
        }
      }
      if (StringUtil.containsChar(" =", context.completionChar)) {
        context.setAddCompletionChar(false)
      }
      EditorModificationUtil.insertStringAtCaret(context.editor, "=")
    }
  }
}
